consensus/hotstuff/signature/weighted_signature_aggregator.go

```diff
@@ -1,39 +1,141 @@
+	// build the internal maps for a faster look-up
+	for i, id := range signers {
+		weightedAgg.idToIndex[id.NodeID] = i
+		weightedAgg.idToWeights[id.NodeID] = id.Stake
+	}
+	return weightedAgg, nil
+}
+
+// Verify verifies the signature under the stored public and message.
 func (s *WeightedSignatureAggregator) Verify(signerID flow.Identifier, sig crypto.Signature) (bool, error) {
-	panic("to be implemented")
+	index, ok := s.idToIndex[signerID]
+	if !ok {
+		return false, fmt.Errorf("couldn't find signerID %s in the index map", signerID)
+	}
+	return s.SignatureAggregatorSameMessage.Verify(index, sig)
 }
 
-// TrustedAdd adds an already verified signature, and look up the weight for the given signer,
-// and add it to the total weight, and returns the total weight that have been collected.
-// return (1000, nil) means the signature has been added, and 1000 weight has been collected in total.
-// return (1000, nil) means the signature is a duplication and 1000 weight has been collected in total.
-func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (totalWeight uint64, exception error) {
-	panic("to be implemented")
+// TrustedAdd adds a signature to the internal set of signatures.
+//
+// It adds the signer's weight to the total collected weight and returns the total weight regardless
+// of the returned error.
+// The function errors if a signature from the signerID was already collected.
+// The function is thread-safe
+func (s *WeightedSignatureAggregator) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (uint64, error) {
+	// get the total weight safely
+	collectedWeight := s.TotalWeight()
+
+	// get the index
+	index, ok := s.idToIndex[signerID]
+	if !ok {
+		return collectedWeight, fmt.Errorf("couldn't find signerID %s in the index map", signerID)
+	}
+	// get the weight
+	weight, ok := s.idToWeights[signerID]
+	if !ok {
+		return collectedWeight, fmt.Errorf("couldn't find signerID %s in the weight map", signerID)
+	}
+
+	// atomically update the signatures pool and the total weight
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// This is a sanity check because the upper layer should have already checked for double-voters.
+	_, ok = s.collectedIDs[signerID]
+	if ok {
+		return collectedWeight, fmt.Errorf("SigneID %s was already added", signerID)
+	}
+
+	err := s.SignatureAggregatorSameMessage.TrustedAdd(index, sig)
+	if err != nil {
+		return collectedWeight, fmt.Errorf("Trusted add has failed: %w", err)
+	}
+
+	s.collectedIDs[signerID] = struct{}{}
+	collectedWeight += weight
+	s.totalWeight = collectedWeight
+	return collectedWeight, nil
 }
 
-// TotalWeight returns the total weight presented by the collected sig shares.
+// TotalWeight returns the total weight presented by the collected signatures.
+//
+// The function is thread-safe
 func (s *WeightedSignatureAggregator) TotalWeight() uint64 {
-	panic("to be implemented")
+	s.lock.RLock()
+	collectedWeight := s.totalWeight
```

Comment:
```
s.lock.Rlock()
defer s.lock.RUnlock()
return s.totalWeight
```

Reply:
Actually we might be even good without locking at all. Since update happens under mutex we will always observe latest value. In other words it will behave like relaxed atomic. What we won't be able to do it is to wait for lock when some other routing is updating weight, but that might be not needed. @AlexHentschel

Reply:
Sorry for the confusing PR, I actually haven't worked on `WeightedSignatureAggregator` yet, this PR is only for `SignatureAggregatorSameMessage`. I am working on `WeightedSignatureAggregator` in another PR. I just updated the description of the current PR (I'll still take your suggestions into account)